src/ray/core_worker/core_worker.cc

```diff
@@ -388,12 +399,115 @@ Status CoreWorker::SerializeActorHandle(const ActorID &actor_id,
   return status;
 }
 
-const ResourceMappingType CoreWorker::GetResourceIDs() const {
-  if (worker_type_ == WorkerType::DRIVER) {
-    ResourceMappingType empty;
-    return empty;
+const ResourceMappingType CoreWorker::GetResourceIDs() const { return resource_ids_; }
+
+void CoreWorker::StartExecutingTasks() {
+  idle_profile_event_.reset(new worker::ProfileEvent(profiler_, "worker_idle"));
+  task_execution_service_->run();
+}
+
+void CoreWorker::StopExecutingTasks() {
+  // Stop main IO service.
+  std::shared_ptr<boost::asio::io_service> task_execution_service =
+      task_execution_service_;
+  // Delay the execution of io_service::stop() to avoid deadlock if
+  // Stop is called inside a task.
+  idle_profile_event_.reset();
+  task_execution_service_->post(
+      [task_execution_service]() { task_execution_service->stop(); });
+}
+
+Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
+                               const ResourceMappingType &resource_ids,
+                               std::vector<std::shared_ptr<RayObject>> *results) {
+  idle_profile_event_.reset();
+  RAY_LOG(DEBUG) << "Executing task " << task_spec.TaskId();
+
+  resource_ids_ = resource_ids;
+  worker_context_.SetCurrentTask(task_spec);
+  SetCurrentTaskId(task_spec.TaskId());
+
+  RayFunction func{task_spec.GetLanguage(), task_spec.FunctionDescriptor()};
+
+  std::vector<std::shared_ptr<RayObject>> args;
+  std::vector<ObjectID> arg_reference_ids;
+  RAY_CHECK_OK(BuildArgsForExecutor(task_spec, &args, &arg_reference_ids));
+
+  std::vector<ObjectID> return_ids;
+  for (size_t i = 0; i < task_spec.NumReturns(); i++) {
+    return_ids.push_back(task_spec.ReturnId(i));
+  }
+
+  Status status;
+  TaskType task_type = TaskType::NORMAL_TASK;
+  if (task_spec.IsActorCreationTask()) {
+    RAY_CHECK(return_ids.size() > 0);
+    return_ids.pop_back();
+    task_type = TaskType::ACTOR_CREATION_TASK;
+    SetActorId(task_spec.ActorCreationId());
+  } else if (task_spec.IsActorTask()) {
+    RAY_CHECK(return_ids.size() > 0);
+    return_ids.pop_back();
+    task_type = TaskType::ACTOR_TASK;
   }
-  return task_execution_interface_->GetResourceIDs();
+  status = task_execution_callback_(task_type, func,
+                                    task_spec.GetRequiredResources().GetResourceMap(),
+                                    args, arg_reference_ids, return_ids, results);
+
+  SetCurrentTaskId(TaskID::Nil());
+  worker_context_.ResetCurrentTask(task_spec);
+
+  // TODO(zhijunfu):
+  // 1. Check and handle failure.
+  // 2. Save or load checkpoint.
+  idle_profile_event_.reset(new worker::ProfileEvent(profiler_, "worker_idle"));
+  return status;
+}
+
+Status CoreWorker::BuildArgsForExecutor(const TaskSpecification &task,
```

Comment:
This looks like it could be a helper function outside the `CoreWorker`.

Reply:
Not quite, because it needs to `Get` objects